File: epi_judge_python/test_framework/random_sequence_checker.py

import collections
import math
import random

from test_framework.test_failure import TestFailure


def compute_deviation_multiplier(allowed_false_negative, num_rvs=1):

    individual_rv_error = allowed_false_negative / num_rvs

    ERROR_BOUNDS = [
        1 - 0.682689492137086, 1 - 0.954499736103642, 1 - 0.997300203936740,
        1 - 0.999936657516334, 1 - 0.999999426696856, 1 - 0.999999998026825,
        1 - 0.999999999997440
    ]

    for i, bound in enumerate(ERROR_BOUNDS):
        if bound <= individual_rv_error:
            return i + 1

    return len(ERROR_BOUNDS) + 1


def check_frequencies(seq, n, false_negative_tolerance):
    avg = len(seq) / float(n)
    k_indiv = compute_deviation_multiplier(false_negative_tolerance, n)
    p = 1.0 / n
    sigma_indiv = math.sqrt(len(seq) * p * (1.0 - p))
    k_sigma_indiv = k_indiv * sigma_indiv
    # To make our testing meaningful "sufficiently large", we need to have enough testing data.
    if len(seq) * p < 50 or len(seq) * (1.0 - p) < 50:
        return True  # Sample size is too small so we cannot use normal approximation

    indiv_freqs = collections.Counter(seq)

    # Check that there are roughly len(seq)/n occurrences of key.
    # By roughly we mean the difference is less than k_sigma.
    return all(
        abs(freq - avg) <= k_sigma_indiv for freq in indiv_freqs.values())


def check_pairs_frequencies(seq, n, false_negative_tolerance):
    seq_pairs = [(x, next_x) for x, next_x in zip(seq, seq[1:])]
    return check_frequencies(seq_pairs, n * n, false_negative_tolerance)


def check_triples_frequencies(seq, n, false_negative_tolerance):
    seq_triples = [(x, next_x, next_next_x)
                   for x, next_x, next_next_x in zip(seq, seq[1:], seq[2:])]
    return check_frequencies(seq_triples, n * n * n, false_negative_tolerance)


def check_birthday_spacings(seq, n):
    expected_avg_repetition_length = int(
        math.ceil(math.sqrt(math.log(2) * 2.0 * n)))
    number_of_subarrays = len(seq) - expected_avg_repetition_length + 1
    MIN_NUMBER_SUBARRAYS = 1000
    if number_of_subarrays < MIN_NUMBER_SUBARRAYS:
        return True  # Not enough subarrays for birthday spacing check

    number_of_subarrays_with_repetitions = sum(
        len(set(seq[i:i + expected_avg_repetition_length])) <
        expected_avg_repetition_length
        for i in range(len(seq) - expected_avg_repetition_length))

    COUNT_TOLERANCE = 0.4
    return COUNT_TOLERANCE * number_of_subarrays <= number_of_subarrays_with_repetitions


# seq is a sequence of integers, which should be in the range [0,n-1]. We assume n << len(seq).
def check_sequence_is_uniformly_random(seq, n, false_negative_tolerance):
    return check_frequencies(seq, n, false_negative_tolerance) and \
           check_pairs_frequencies(seq, n, false_negative_tolerance) and \
           check_triples_frequencies(seq, n, false_negative_tolerance) and \
           check_birthday_spacings(seq, n)


def binomial_coefficient(n, k):
    return 0 if n < k else math.factorial(n) // (
        math.factorial(k) * math.factorial(n - k))


# Get the mth combination in lexicographical order from A (n elements) chosen k at a time.
def compute_combination_idx(A, n, k, m):
    comb = []
    a, b, x = n, k, (binomial_coefficient(n, k) - 1) - m
    for _ in range(k):
        a = a - 1
        while binomial_coefficient(a, b) > x:
            a -= 1
        comb.append(A[n - 1 - a])
        x -= binomial_coefficient(a, b)
        b -= 1
    return comb


# Run func several times, and early return if it succeeds; otherwise, throws TestFailure.
def run_func_with_retries(func):
    NUM_RUNS = 5
    # If any of the run succeed, we assume the algorithm is correct.
    if any(func() for _ in range(NUM_RUNS)):
        return
    raise TestFailure("Result is not a random permutation")


if __name__ == '__main__':
    assert check_sequence_is_uniformly_random(
        [random.randint(1, 10) for i in range(1000000)],
        10,
        false_negative_tolerance=0.01)
    assert not check_sequence_is_uniformly_random(
        [random.randint(1, 11) for i in range(1000000)],
        10,
        false_negative_tolerance=0.01)
    assert not check_sequence_is_uniformly_random(
        [i % 10 for i in range(1000000)], 10, false_negative_tolerance=0.01)

File: epi_judge_java_solutions/epi/test_framework/RandomSequenceChecker.java

package epi.test_framework;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;
import java.util.concurrent.Callable;

public class RandomSequenceChecker {
  private static int computeDeviationMultiplier(double allowedFalseNegative,
                                                int numRvs) {
    final double individualRvError = allowedFalseNegative / numRvs;
    final double[] ERROR_BOUNDS = {1 - 0.682689492137086, 1 - 0.954499736103642,
                                   1 - 0.997300203936740, 1 - 0.999936657516334,
                                   1 - 0.999999426696856, 1 - 0.999999998026825,
                                   1 - 0.999999999997440};

    for (int i = 0; i < ERROR_BOUNDS.length; ++i) {
      if (ERROR_BOUNDS[i] <= individualRvError) {
        return i + 1;
      }
    }
    return ERROR_BOUNDS.length + 1;
  }

  private static boolean checkFrequencies(List<Integer> seq, int n,
                                          double falseNegativeTolerance) {
    final double avg = (double)seq.size() / n;
    final int kIndiv = computeDeviationMultiplier(falseNegativeTolerance, n);
    final double p = 1.0 / n;
    final double sigmaIndiv = Math.sqrt(seq.size() * p * (1.0 - p));
    final double kSigmaIndiv = kIndiv * sigmaIndiv;
    // To make our testing meaningful "sufficiently large", we need to have
    // enough testing data.
    if (seq.size() * p < 50 || seq.size() * (1 - p) < 50) {
      return true; // Sample size is too small so we cannot use normal
                   // approximation
    }

    Map<Integer, Integer> indivFreqs = new HashMap<>();
    for (Integer a : seq) {
      indivFreqs.put(a, indivFreqs.getOrDefault(a, 0) + 1);
    }

    // Check that there are roughly seq.size()/n occurrences of key. By roughly
    // we mean the difference is less than k_sigma.
    return indivFreqs.values().stream().allMatch(
        freq -> Math.abs(avg - freq) <= kSigmaIndiv);
  }

  private static boolean checkPairsFrequencies(List<Integer> seq, int n,
                                               double falseNegativeTolerance) {
    List<Integer> seqPairs = new ArrayList<>();
    for (int i = 1; i < seq.size(); ++i) {
      seqPairs.add(seq.get(i - 1) * n + seq.get(i));
    }

    return checkFrequencies(seqPairs, n * n, falseNegativeTolerance);
  }

  private static boolean
  checkTriplesFrequencies(List<Integer> seq, int n,
                          double falseNegativeTolerance) {
    List<Integer> seqTriples = new ArrayList<>();
    for (int i = 2; i < seq.size(); ++i) {
      seqTriples.add(seq.get(i - 2) * n * n + seq.get(i - 1) * n + seq.get(i));
    }

    return checkFrequencies(seqTriples, n * n * n, falseNegativeTolerance);
  }

  private static boolean checkBirthdaySpacings(List<Integer> seq, int n) {
    final int expectedAvgRepetitionLength =
        (int)Math.ceil(Math.sqrt(Math.log(2.0) * 2.0 * n));
    final int numberOfSubarrays = seq.size() - expectedAvgRepetitionLength + 1;
    final int MIN_NUMBER_SUBARRAYS = 1000;
    if (numberOfSubarrays < MIN_NUMBER_SUBARRAYS) {
      return true; // Not enough subarrays for birthday spacing check
    }

    int numberOfSubarraysWithRepetitions = 0;
    for (int i = 0; i < seq.size() - expectedAvgRepetitionLength; ++i) {
      Set<Integer> seqWindow =
          new HashSet<>(seq.subList(i, i + expectedAvgRepetitionLength));
      numberOfSubarraysWithRepetitions +=
          (seqWindow.size() < expectedAvgRepetitionLength) ? 1 : 0;
    }

    final double COUNT_TOLERANCE = 0.4;
    return COUNT_TOLERANCE * numberOfSubarrays <=
        numberOfSubarraysWithRepetitions;
  }

  // seq is a sequence of integers, which should be in the range [0,n-1]. We
  // assume n << seq.size().
  public static boolean
  checkSequenceIsUniformlyRandom(List<Integer> seq, int n,
                                 double falseNegativeTolerance) {
    return checkFrequencies(seq, n, falseNegativeTolerance) &&
        checkPairsFrequencies(seq, n, falseNegativeTolerance) &&
        checkTriplesFrequencies(seq, n, falseNegativeTolerance) &&
        checkBirthdaySpacings(seq, n);
  }

  public static int binomialCoefficient(int n, int k) {
    if (n < k) {
      return 0;
    }
    if (k == 0 || k == n) {
      return 1;
    }
    return binomialCoefficient(n - 1, k - 1) + binomialCoefficient(n - 1, k);
  }

  // Get the mth combination in lexicographical order from A (n elements) chosen
  // k at a time.
  public static List<Integer> computeCombinationIdx(List<Integer> A, int n,
                                                    int k, int m) {
    List<Integer> comb = new ArrayList<>();
    int a = n, b = k, x = (binomialCoefficient(n, k) - 1) - m;
    for (int i = 0; i < k; ++i) {
      --a;
      while (binomialCoefficient(a, b) > x) {
        --a;
      }
      comb.add(A.get(n - 1 - a));
      x -= binomialCoefficient(a, b);
      --b;
    }
    return comb;
  }

  public static void runFuncWithRetries(Callable<Boolean> func)
      throws Exception {
    final int NUM_RUNS = 5;
    for (int i = 0; i < NUM_RUNS; i++) {
      if (func.call()) {
        return;
      }
    }
    throw new TestFailure("Result is not a random permutation");
  }

  public static void main(String[] args) {
    Random gen = new Random();
    List<Integer> seq = new ArrayList<>();
    for (int i = 0; i < 1000000; ++i) {
      seq.add(gen.nextInt(10) + 1);
    }
    System.out.println(checkSequenceIsUniformlyRandom(seq, 10, 0.01));

    seq.clear();
    for (int i = 0; i < 1000000; ++i) {
      seq.add(gen.nextInt(11) + 1);
    }
    System.out.println(checkSequenceIsUniformlyRandom(seq, 10, 0.01));

    seq.clear();
    for (int i = 0; i < 1000000; ++i) {
      seq.add(i % 10);
    }
    System.out.println(checkSequenceIsUniformlyRandom(seq, 10, 0.01));
  }
}
